source/common/router/req_header_formatter.cc

```diff
@@ -0,0 +1,99 @@
+#include "common/router/req_header_formatter.h"
+
+#include <cstdint>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/http/headers.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+HeaderFormatterSharedPtr RequestHeaderParser::parseInternal(const std::string& format) {
+  std::string variable_name;
+  if (format.find("%") == 0) {
+    size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format(
+          "Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",
+          format));
+    }
+    variable_name = format.substr(1, last_occ_pos - 1);
+
+  } else {
+    HeaderFormatterSharedPtr plain_header_formatter_shard_ptr(new PlainHeaderFormatter(format));
+    return plain_header_formatter_shard_ptr;
+  }
+  HeaderFormatterSharedPtr request_header_formatter_shard_ptr(
+      new RequestHeaderFormatter(variable_name));
+  return request_header_formatter_shard_ptr;
+}
+
+RequestHeaderParserSharedPtr RequestHeaderParser::parse(const Json::Object& config) {
+
+  RequestHeaderParserSharedPtr request_header_parser(new RequestHeaderParser());
+  if (config.hasObject("request_headers_to_add")) {
+    std::vector<Json::ObjectSharedPtr> request_headers =
+        config.getObjectArray("request_headers_to_add");
+    for (const Json::ObjectSharedPtr& header : request_headers) {
+      if (!header->getString("key").empty() && !header->getString("value").empty()) {
+        log().debug("adding key {} to header formatter map", header->getString("key"));
+        request_header_parser->header_formatter_map_.emplace(
+            std::make_pair(Http::LowerCaseString(header->getString("key")),
+                           RequestHeaderParser::parseInternal(header->getString("value"))));
+      }
+    }
+  }
+
+  return request_header_parser;
+}
+
+void RequestHeaderParser::evaluateRequestHeaders(
+    Http::HeaderMap& headers, Http::AccessLog::RequestInfo& requestInfo,
+    const std::list<std::pair<Http::LowerCaseString, std::string>>& requestHeadersToAdd) const {
+
+  for (const std::pair<Http::LowerCaseString, std::string>& to_add : requestHeadersToAdd) {
+
+    log().debug("request headers key {}", to_add.first.get());
+
+    // search if the request has formatters defined
+    if (header_formatter_map_.count(to_add.first) == 1) {
+
+      auto search = header_formatter_map_.find(to_add.first);
+      // access the formatter
+      if (search != header_formatter_map_.end()) {
```

Comment:
There shouldn't be an `if` here, you tested the condition in the outer `if`. I would rewrite it to do the `find`, and then just condition on if `!= ..end()` on the outer `if`.